crates/mipsy_lib/src/inst/instruction.rs

```diff
@@ -525,45 +574,58 @@ impl ArgumentType {
-                            matches!(self, Self::J | Self::U32 | Self::Off32Rs | Self::Off32Rt)
-                        }
-                        MpImmediate::LabelReference(_) => match self {
-                            Self::I32 | Self::U32 | Self::J | Self::Off32Rs | Self::Off32Rt => true,
-                            Self::I16 => relative_label,
-                            _ => false,
-                        },
+            MpArgument::Number(number) => match number {
+                MpNumber::Immediate(immediate) => match immediate {
+                    &MpImmediate::I16(num) => match self {
+                        Self::I16 | Self::I32 | Self::Off32Rs | Self::Off32Rt => true,
+                        Self::U16 | Self::U32 => num >= 0,
+                        Self::Shamt => (0..=31).contains(&num),
+                        _ => false,
                     },
-                    MpNumber::BinaryOpImmediate(_imm1, _op, _imm2) => {
-                        // TODO(zkol): this is brittle and based on faulty assumptions
-                        matches!(self, Self::I32 | Self::U32 | Self::Off32Rs | Self::Off32Rt)
-                    }
-                    MpNumber::Char(_) => {
-                        matches!(self, Self::I16 | Self::I32 | Self::U16 | Self::U32)
+                    MpImmediate::U16(_) => matches!(
+                        self,
+                        Self::U16 | Self::I32 | Self::U32 | Self::Off32Rs | Self::Off32Rt
+                    ),
+                    &MpImmediate::I32(num) => match self {
+                        Self::I32 | Self::J | Self::Off32Rs | Self::Off32Rt => true,
+                        Self::U32 => num >= 0,
+                        _ => false,
+                    },
+                    MpImmediate::U32(_) => {
+                        matches!(self, Self::J | Self::U32 | Self::Off32Rs | Self::Off32Rt)
                     }
-                    MpNumber::Float32(_) => matches!(self, Self::F32 | Self::F64),
-                    MpNumber::Float64(_) => matches!(self, Self::F64),
+                    MpImmediate::LabelReference(_) => match self {
+                        Self::I32 | Self::U32 | Self::J | Self::Off32Rs | Self::Off32Rt => true,
+                        Self::I16 => relative_label,
+                        _ => false,
+                    },
+                },
+                MpNumber::Constant(cnst) => eval_constant(program, cnst, "".into())
+                    // skip over the unresolved constant error here
+                    // since immediate constants wouldnt be defined
+                    .or_else(|e| {
+                        if let crate::MipsyError::Compiler(c) = &e {
+                            if matches!(c.error(), UnresolvedConstant { .. }) {
+                                return Ok(0);
+                            }
+                        }
+                        return Err(e);
+                    })
+                    .map_err(InternalError::from)
```

Comment:
this loses error context which is needed now that we propagate errors from here. in this specific case it means the `^^^^^^` indicator thing will underline the entire line instead of just the focal constant